modules/nextflow/src/main/groovy/nextflow/script/OutputDsl.groovy

```diff
@@ -0,0 +1,250 @@
+    }
+
+    void tags(Map value) {
+        setDefault('tags', value)
+    }
+
+    private void setDefault(String name, Object value) {
+        if( defaults.containsKey(name) )
+            throw new ScriptRuntimeException("Default `${name}` option cannot be defined more than once in the workflow publish definition")
+        defaults[name] = value
+    }
+
+    void target(String name, Closure closure) {
+        if( publishConfigs.containsKey(name) )
+            throw new ScriptRuntimeException("Target '${name}' is defined more than once in the workflow publish definition")
+
+        final dsl = new TargetDsl()
+        final cl = (Closure)closure.clone()
+        cl.setResolveStrategy(Closure.DELEGATE_FIRST)
+        cl.setDelegate(dsl)
+        cl.call()
+
+        publishConfigs[name] = dsl.getOptions()
+    }
+
+    void build(Map<DataflowWriteChannel,String> targets) {
+        // construct mapping of target name -> source channels
+        final Map<String,List<DataflowWriteChannel>> publishSources = [:]
+        for( final source : targets.keySet() ) {
+            final name = targets[source]
+            if( !name )
+                continue
+            if( name !in publishSources )
+                publishSources[name] = []
+            publishSources[name] << source
+        }
+
+        // create publish op (and optional index op) for each target
+        for( final name : publishSources.keySet() ) {
+            final sources = publishSources[name]
+            final mixed = sources.size() > 1
+                ? new MixOp(sources.collect( ch -> CH.getReadChannel(ch) )).apply()
+                : sources.first()
+            final opts = publishOptions(name, publishConfigs[name] ?: [:])
+
+            new PublishOp(CH.getReadChannel(mixed), opts).apply()
+
+            if( opts.index ) {
+                final basePath = opts.path as Path
+                final indexOpts = opts.index as Map
+                final indexPath = indexOpts.path as String
+                if( !indexPath )
+                    throw new ScriptRuntimeException("Index file definition for publish target '${name}' is missing `path` option")
+                new PublishIndexOp(CH.getReadChannel(mixed), basePath, indexPath, indexOpts).apply()
+            }
+        }
+    }
+
+    private Map publishOptions(String name, Map overrides) {
+        if( !directory )
+            directory = Paths.get('.').complete()
+
+        final opts = defaults + overrides
+        if( opts.containsKey('ignoreErrors') )
+            opts.failOnError = !opts.remove('ignoreErrors')
+        if( !opts.containsKey('overwrite') )
+            opts.overwrite = 'standard'
+
+        final path = opts.path as String ?: name
+        if( path.startsWith('/') )
+            throw new ScriptRuntimeException("Invalid publish target path '${path}' -- it should be a relative path")
+        opts.path = directory.resolve(path)
+        return opts
+    }
+
+    static class TargetDsl {
```

Comment:
Not 100% sure, but likely inner class should be annotated `@CompileStatic`

Reply:
Since I've been digging though the groovy codebase... I can confirm it is not needed 😄 

https://github.com/apache/groovy/blob/55d43696d2b13fa03067ef3c2b200eb7c84148ca/src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java#L450-L451